### DISCUSSION TITLE
Best practice when upstream image contains a security issue

### DISCUSSION THREAD

**Discussion by bmauter** [BretFisher/ama#135] [Q&A]
I was hoping you could comment on some best practices on how to handle updating containers once they're in production.  For example, let's pretend a new nasty kernel bug came out and I have a couple dozen containers that are affected.  Once the OS image that I built from has a new image with the updated kernel in it, now I probably have to rebuild all of my images and deploy.  How do you recommend handling that?

If each image is made up of layers, can the base layer be replaced and the container restarted without having to rebuild anything else?  I thought each layer was hashed so that Docker would know if something changed.  If the project's Docker build files are available, I guess a rebuild of the final image would be easier than trying to trick it to use an updated layer.

What if there's a different development team affected by each of those affected containers?  Should we expect everyone to drop what they're doing and rebuild their images and redeploy?  Should we just designate one person to rebuild them all?

Maybe our vulnerability checker should trigger our CI/CD process to just rebuild and redeploy things as needed.

Won't QA departments want a full regression test, regardless of who does it and how the new image was built?  It's a new image, after all.

Thanks,

-Brian

**Comment by BretFisher** [comment]
Generally, you are:
1. Rebuilding images in a CI solution when any code needs to change. That may be your own commits, that maybe the app (say mysql) version needs to change, *or* that maybe that an application dependency (say openssl). Swapping out layers isn't a common practice yet. 
2. One of the reasons you wouldn't want to swap out a FROM image, is for ensuring reliable testing. The more automated you get, with workflows like "commit to master branch will autobuild and test in CI, and then CD to production infrastructure". If you were to swap out a layer or a FROM image somehow, you'd still want to "build" the layers again to ensure they work after those lower layer changes.
3. In my [dockercon19 talk and examples](https://github.com/BretFisher/dockercon19), which happened to be Node.js specific, is a common pattern for any framework or language to break up the parts of a CI workflow into a single Dockerfile, so that yes, you're testing, auditing, scanning, and then copying code to a minimal image for prod deployment. I really like sticking as much of this as I can into the docker and compose files, because they can run anyway. You want to run the CI workflow locally, no problem... just `docker-compose up --build` and maybe a few more `docker-compose run` commands for test runners. 
4. No humans should be locally building images that get pushed to servers. All of that work is done by CI based on code commits. 
4. If you have to wait on humans to test code changes before it's deployed, then that's the area to focus on improving, by removing the humans from the process. Expecting humans to test every release is a key limiter to agility, and would be a big red flag for any project where I'm tasked to improve DevOps and automation. Replace the human tests with code. Turn the Q/A dept into people writing automated tests that will run during container build, and improving operations testing so you'll know quicker when problems arise after deployment :)
3. In any project, I guide the evolution of the workflow and automation towards a few main goals:
  - Testing can easily be run anywhere. On a dev's local machine or by CI, without a special CI solution. You'll notice all the newest CI products on the market that are container focused, are quite simple compared to the old solutions. They have a lot less to do because we've moved a lot of the work into docker.
  - Ideally, anything required by CI can be created locally by docker-compose, so dev's don't have to be locked into only using a remote CI solution that's treated as a magic box.
  - Every change of all (or maybe just important) branches gets tested, meaning images are rebuilt. That way no "little" change sneaks into production without full testing. Many production failures I've seen are caused by "innocent little changes that wouldn't affect anything" 🤣
  - Usually, if all that needed to be improved, then ops monitoring and rollback routines need to be improved as well. You can't increase the speed of deployment without improving the ability to monitor that and recover from outages.
You'll notice that a lot of this is "humans changing" rather than tools changing. 
I hope that helps.
